label_studio/fsm/registry.py

```diff
@@ -0,0 +1,420 @@
+                return denormalizer(entity)
+            except Exception as e:
+                logger.error(f'Error getting denormalized fields for {entity_name}: {e}')
+
+        return {}
+
+    def is_registered(self, entity_name: str) -> bool:
+        """Check if a model is registered for an entity type."""
+        return entity_name.lower() in self._models
+
+    def clear(self):
+        """Clear all registered models (useful for testing)."""
+        self._models.clear()
+        self._denormalizers.clear()
+        self._initialized = False
+        logger.debug('Cleared state model registry')
+
+    def get_all_models(self) -> Dict[str, 'BaseState']:
+        """Get all registered models."""
+        return self._models.copy()
+
+
+# Global registry instance
+state_model_registry = StateModelRegistry()
+
+
+def register_state_model(entity_name: str, denormalizer: Optional[Callable[[Model], Dict[str, Any]]] = None):
+    """
+    Decorator to register a state model.
+
+    Args:
+        entity_name: Name of the entity (e.g., 'task', 'annotation')
+        denormalizer: Optional function to extract denormalized fields
+
+    Example:
+        @register_state_model('task')
+        class TaskState(BaseState):
+            # ... implementation
+    """
+
+    def decorator(state_model: 'BaseState') -> 'BaseState':
+        state_model_registry.register_model(entity_name, state_model, denormalizer)
+        return state_model
+
+    return decorator
+
+
+def register_state_model_class(
+    entity_name: str, state_model: 'BaseState', denormalizer: Optional[Callable[[Model], Dict[str, Any]]] = None
+):
+    """
+    Convenience function to register a state model programmatically.
+
+    Args:
+        entity_name: Name of the entity (e.g., 'task', 'annotation')
+        state_model: The state model class for this entity
+        denormalizer: Optional function to extract denormalized fields
+    """
+    state_model_registry.register_model(entity_name, state_model, denormalizer)
+
+
+def get_state_model(entity_name: str) -> Optional['BaseState']:
+    """
+    Convenience function to get a state model.
+
+    Args:
+        entity_name: Name of the entity
+
+    Returns:
+        State model class or None if not registered
+    """
+    return state_model_registry.get_model(entity_name)
+
+
+def get_state_model_for_entity(entity: Model) -> Optional['BaseState']:
+    """Get the state model for an entity."""
+    entity_name = entity._meta.model_name.lower()
+    return get_state_model(entity_name)
+
+
+class TransitionRegistry:
+    """
+    Registry for managing declarative transitions.
+
+    Provides a centralized way to register, discover, and execute transitions
+    for different entity types and state models.
+    """
+
+    def __init__(self):
+        self._transitions: Dict[str, Dict[str, 'BaseTransition']] = {}
+
+    def register(self, entity_name: str, transition_name: str, transition_class: 'BaseTransition'):
```

Comment:
just curious does the system allow to register multiple transitions, with the same current state and target state?

Reply:
How is it enforced then, that we cannot register multiple transitions with the same state?